monads/src/HW/Eval.hs

```diff
@@ -29,8 +29,41 @@ initialState = MachineState [] M.empty
 -- Execute a single instruction. 
 -- Successful evaluation does not produce any useful result: only the effect of modifying state matters. 
 execInstr :: (Ord v, Show v) => StackInstr v -> MyState (MachineState v) (Either (Error v) ())
-execInstr = undefined 
+execInstr (PushNum x) = do
+  modify $ \s -> s { getStack = x : getStack s }
+  return $ Right ()
+execInstr (PushVar k) = do
+  env <- gets getEnv
+  case M.lookup k env of
+    Just v -> do
+      modify $ \s -> s { getStack = v : getStack s }
+      return $ Right ()
+    Nothing -> return $ Left (VarUndefined k)
+execInstr Add = do
+  stack <- gets getStack
+  case stack of
+    (a:b:_) -> do
+      modify $ \s -> s { getStack = a + b : drop 2 (getStack s)}
```

Comment:
`drop 2` and `getStack s` are excessive, since you can just get this value from the pattern match in line 45.